cip/1.accepted/CIP2016-12-14-Constraint-syntax.adoc

```diff
@@ -0,0 +1,285 @@
+CREATE (:Color {name: 'black', rgb: 0x000000})
+CREATE (:Color {name: 'very, very dark grey', rgb: 0x000000}) // rounding error!
+----
+
+Due to the duplication of the `rgb` property, the following attempt at creating a constraint will fail:
+
+[source, cypher]
+----
+CREATE CONSTRAINT only_one_color_per_rgb
+FOR (c:Color)
+REQUIRE UNIQUE c.rgb
+----
+
+Suppose that we would rather like to have one color node per name _and_ RGB value (to work around the rounding errors).
+We could then use the following constraint, without modifying our data:
+
+[source, cypher]
+----
+CREATE CONSTRAINT unique_color_nodes
+FOR (c:Color)
+REQUIRE UNIQUE c.rgb, c.name
+----
+
+[[existence]]
+=== Property existence constraints
+
+Property existence constraints are defined for both nodes and relationships, but the semantics are the same.
+For this reason we will go over both constraints in the same section.
+
+==== Syntax
+
+.Grammar definition for property existence constraint:
+[source, ebnf]
+----
+existence-constraint = "exists", "(", property-expression, ")" ;
+----
+
+.Example of node property existence constraint:
+[source, cypher]
+----
+CREATE CONSTRAINT colors_must_have_rgb
+FOR (c:Color)
+REQUIRE exists(c.rgb)
+----
+
+.Example of relationship property existence constraint:
+[source, cypher]
+----
+CREATE CONSTRAINT rates_have_quality
+FOR ()-[l:RATED]-()
+REQUIRE exists(l.rating)
+----
+
+==== Semantics
+
+The domain of a node property existence constraint are all nodes with the specified label.
+Similarly, the domain of a relationship property existence constraint are all relationship with the specified type.
+
+Property existence constraints mandates that the value of the specified property exists (is non-null) for all entities in the domain.
+
+===== Example
+
+Consider the graph containing `:Color` nodes.
+Each color has an integral RGB value representation in a property `rgb`.
+Users may lookup color nodes to extract their RGB values for application processing.
+Users may also add new color nodes to the graph.
+
+Suppose the query that looks up the RGB value of a color with a given name looks like this:
+
+[source, cypher]
+----
+MATCH (c:Color {name: $name})
+WHERE exists(c.rgb)
+RETURN c.rgb
+----
+
+The `WHERE` clause protects the application from receiving `null` values back for user-defined colors where the RGB values have not been specified correctly.
+It may however be eliminated by the introduction of a node property existence constraint:
+
+[source, cypher]
+----
+CREATE CONSTRAINT colors_must_have_rgb
+FOR (c:Color)
+REQUIRE exists(c.rgb)
+----
+
+Any updating statement that would create a `:Color` node without specifying a `rgb` property for it would now fail.
+
+=== Interaction with existing features
+
+The main interaction between the constraints and the rest of the language happens during updating statements.
+Existing constraints will cause certain updating statements to fail; in fact, that's the main purpose.
```

Comment:
"; in fact, that's the main purpose"
->
" - which is, in fact, the main purpose of this feature"